staging/src/k8s.io/csi-translation-lib/plugins/in_tree_volume.go

```diff
@@ -124,38 +172,54 @@ func addTopology(pv *v1.PersistentVolume, topologyKey string, zones []string) er
 	}
 
 	// Make sure the necessary fields exist
-	pv.Spec.NodeAffinity = new(v1.VolumeNodeAffinity)
-	pv.Spec.NodeAffinity.Required = new(v1.NodeSelector)
-	pv.Spec.NodeAffinity.Required.NodeSelectorTerms = make([]v1.NodeSelectorTerm, 1)
+	if pv.Spec.NodeAffinity == nil {
+		pv.Spec.NodeAffinity = new(v1.VolumeNodeAffinity)
+	}
+
+	if pv.Spec.NodeAffinity.Required == nil {
+		pv.Spec.NodeAffinity.Required = new(v1.NodeSelector)
+	}
+
+	if len(pv.Spec.NodeAffinity.Required.NodeSelectorTerms) == 0 {
+		pv.Spec.NodeAffinity.Required.NodeSelectorTerms = make([]v1.NodeSelectorTerm, 1)
+	}
 
 	topology := v1.NodeSelectorRequirement{
 		Key:      topologyKey,
 		Operator: v1.NodeSelectorOpIn,
 		Values:   zones,
 	}
 
-	pv.Spec.NodeAffinity.Required.NodeSelectorTerms[0].MatchExpressions = append(
-		pv.Spec.NodeAffinity.Required.NodeSelectorTerms[0].MatchExpressions,
-		topology,
-	)
+	// add the CSI topology to each term
+	for i := range pv.Spec.NodeAffinity.Required.NodeSelectorTerms {
+		pv.Spec.NodeAffinity.Required.NodeSelectorTerms[i].MatchExpressions = append(
+			pv.Spec.NodeAffinity.Required.NodeSelectorTerms[i].MatchExpressions,
+			topology,
+		)
+	}
 
 	return nil
 }
 
-// translateTopology converts existing zone labels or in-tree topology to CSI topology.
+// translateTopologyFromInTreeToCSI converts existing zone labels or in-tree topology to CSI topology.
 // In-tree topology has precedence over zone labels.
-func translateTopology(pv *v1.PersistentVolume, topologyKey string) error {
+func translateTopologyFromInTreeToCSI(pv *v1.PersistentVolume, topologyKey string) error {
 	// If topology is already set, assume the content is accurate
-	if len(getTopologyZones(pv, topologyKey)) > 0 {
+	if len(getTopologyValues(pv, topologyKey)) > 0 {
 		return nil
 	}
 
-	zones := getTopologyZones(pv, v1.LabelFailureDomainBetaZone)
+	_, zoneLabel, _ := getTopologyLabel(pv)
```

Comment:
Should region keys be removed from CSI NodeAffinity?

Reply:
Good question. I did not change much on the `translateTopologyFromInTreeToCSI` case. So the original behavior is to keep the region keys. @msau42 What do you think. Should we remove it when in-tree -> CSI? Well, actually in the current implementation, the kubernetes zone topology labels is not guaranteed to be removed because what this function does is just to make sure the CSITopologyKey exists, and it dont really care about Kubernetes topology.

Reply:
When is intree -> csi called, and when would it matter? My initial hunch is the scheduler. And perhaps that might cause issues because a PV could have beta region label on it, which would fail scheduling.

Reply:
Maybe we could have a region handler here as well, in case a storage provider has special region support?

Reply:
I think msau mentioned a good point. Suppose that we are migrating an old PV to a new cluster where nodes only have GA region labels. And that might fail the scheduling. So I think the right thing to do here is to remove the beta region label at least/replace the beta region label to GA label as we are sure all the node should have GA label at least.